src/Storages/StorageS3.cpp

```diff
@@ -1091,15 +1193,34 @@ Pipe StorageS3::read(
     size_t max_block_size,
     size_t num_streams)
 {
-    auto query_configuration = updateConfigurationAndGetCopy(local_context);
+    auto read_from_format_info = prepareReadingFromFormat(column_names, storage_snapshot, supportsSubsetOfColumns(local_context), virtual_columns);
```

Comment:
This `read_from_format_info` is used only for getting `source_header`, and then we call `prepareReadingFromFormat` again in `ReadFromStorageS3Step::initializePipeline`. Why don't just pass `read_from_format_info` as an argument in `ReadFromStorageS3Step`?